plugins/experimental/ts_lua/ts_lua_transform.c

```diff
@@ -140,30 +162,38 @@ ts_lua_transform_handler(TSCont contp, ts_lua_http_transform_ctx *transform_ctx,
       n = 2;
     }
 
-    if (towrite == 0) {
+    if (towrite == 0 && empty_input == 0) {
       break;
     }
 
-    blk   = TSIOBufferReaderStart(transform_ctx->reserved.reader);
-    start = TSIOBufferBlockReadStart(blk, transform_ctx->reserved.reader, &blk_len);
-
-    lua_pushlightuserdata(L, transform_ctx);
-    lua_rawget(L, LUA_GLOBALSINDEX); /* push function */
+    if (empty_input == 0) {
+      blk   = TSIOBufferReaderStart(transform_ctx->reserved.reader);
+      start = TSIOBufferBlockReadStart(blk, transform_ctx->reserved.reader, &blk_len);
+
+      lua_pushlightuserdata(L, transform_ctx);
+      lua_rawget(L, LUA_GLOBALSINDEX); /* push function */
+
+      if (towrite > blk_len) {
```

Comment:
Could you clip `towrite` to `blk_len` here and avoid the more complex `if`? You might be able to blend the `empty_input` case in as well, since `towrite == 0` and `eos = 1` in that case already, you just need to avoid touching `reserved.reader`. Even if `start` is `NULL`, it shouldn't matter because the length is `0` (although `start` could be initialized to `""` just as easily).